File: 2048GAME/csharp/Game.cs
// Game.cs — GameBoard class: core 2048 logic (C#)
using System;
using System.Collections.Generic;

namespace Game2048
{
    public class GameBoard
    {
        public const int Size = 4;
        public int[,] Grid { get; private set; } = new int[Size, Size];
        public int Score { get; set; }
        public bool Over { get; set; }
        public bool Won  { get; set; }

        private static readonly Random Rng = new();

        public GameBoard() => Reset();

        public void Reset()
        {
            Grid  = new int[Size, Size];
            Score = 0;
            Over  = false;
            Won   = false;
            AddRandom();
            AddRandom();
        }

        public GameBoard Clone()
        {
            var b = new GameBoard { Score = Score, Over = Over, Won = Won };
            // skip Reset's two AddRandom calls by directly cloning:
            b.Grid = (int[,])Grid.Clone();
            return b;
        }

        // ── Cells ──────────────────────────────────────────────────────────
        public List<(int r, int c)> EmptyCells()
        {
            var cells = new List<(int, int)>();
            for (int r = 0; r < Size; r++)
                for (int c = 0; c < Size; c++)
                    if (Grid[r, c] == 0) cells.Add((r, c));
            return cells;
        }

        public bool AddRandom()
        {
            var cells = EmptyCells();
            if (cells.Count == 0) return false;
            var (r, c) = cells[Rng.Next(cells.Count)];
            Grid[r, c] = Rng.NextDouble() < 0.9 ? 2 : 4;
            return true;
        }

        // ── Row slide ──────────────────────────────────────────────────────
        private static (int[] row, int gained) SlideRow(int[] row)
        {
            var arr = new List<int>();
            foreach (int v in row) if (v != 0) arr.Add(v);
            int gained = 0;
            for (int i = 0; i < arr.Count - 1; i++)
            {
                if (arr[i] == arr[i + 1])
                {
                    arr[i] *= 2;
                    gained += arr[i];
                    arr.RemoveAt(i + 1);
                }
            }
            while (arr.Count < Size) arr.Add(0);
            return (arr.ToArray(), gained);
        }

        // ── Rotate 90° clockwise ──────────────────────────────────────────
        private static int[,] RotateRight(int[,] g)
        {
            var o = new int[Size, Size];
            for (int r = 0; r < Size; r++)
                for (int c = 0; c < Size; c++)
                    o[c, Size - 1 - r] = g[r, c];
            return o;
        }

        // ── Move ──────────────────────────────────────────────────────────
        public (bool moved, int gained) Move(int dir, bool sim = false)
        {
            int[] rotsBefore = { 0, 2, 3, 1 };
            int[] rotsAfter  = { 0, 2, 1, 3 };
            var g = (int[,])Grid.Clone();
            for (int i = 0; i < rotsBefore[dir]; i++) g = RotateRight(g);

            bool moved = false;
            int totalGained = 0;
            for (int r = 0; r < Size; r++)
            {
                var rowIn = new int[Size];
                for (int c = 0; c < Size; c++) rowIn[c] = g[r, c];
                var (newRow, gained) = SlideRow(rowIn);
                bool changed = false;
                for (int c = 0; c < Size; c++)
                    if (newRow[c] != g[r, c]) { changed = true; break; }
                if (changed) moved = true;
                for (int c = 0; c < Size; c++) g[r, c] = newRow[c];
                totalGained += gained;
            }

            for (int i = 0; i < rotsAfter[dir]; i++) g = RotateRight(g);

            if (moved)
            {
                Grid   = g;
                Score += totalGained;
                if (!sim)
                {
                    if (!Won && MaxTile() >= 2048) Won = true;
                    AddRandom();
                    if (!CanMove()) Over = true;
                }
            }
            return (moved, totalGained);
        }

        // ── Utilities ─────────────────────────────────────────────────────
        public bool CanMove()
        {
            if (EmptyCells().Count > 0) return true;
            for (int r = 0; r < Size; r++)
                for (int c = 0; c < Size; c++)
                {
                    int v = Grid[r, c];
                    if (c < Size - 1 && Grid[r, c + 1] == v) return true;
                    if (r < Size - 1 && Grid[r + 1, c] == v) return true;
                }
            return false;
        }

        public int MaxTile()
        {
            int m = 0;
            foreach (int v in Grid) m = Math.Max(m, v);
            return m;
        }

        public List<int> ValidMoves()
        {
            var moves = new List<int>();
            for (int d = 0; d < 4; d++)
            {
                var b = Clone();
                if (b.Move(d, sim: true).moved) moves.Add(d);
            }
            return moves;
        }
    }
}

File: 2048GAME/python/game.py
"""
game.py — GameBoard class: core 2048 logic (Python)
"""
import random
import math
import copy


class GameBoard:
    SIZE = 4

    def __init__(self):
        self.grid = [[0] * 4 for _ in range(4)]
        self.score = 0
        self.over = False
        self.won = False
        self.add_random()
        self.add_random()

    def clone(self):
        b = object.__new__(GameBoard)
        b.grid = [row[:] for row in self.grid]
        b.score = self.score
        b.over = self.over
        b.won = self.won
        return b

    def empty_cells(self):
        return [(r, c) for r in range(4) for c in range(4) if self.grid[r][c] == 0]

    def add_random(self):
        cells = self.empty_cells()
        if not cells:
            return False
        r, c = random.choice(cells)
        self.grid[r][c] = 2 if random.random() < 0.9 else 4
        return True

    @staticmethod
    def _slide_row(row):
        arr = [v for v in row if v != 0]
        gained = 0
        i = 0
        while i < len(arr) - 1:
            if arr[i] == arr[i + 1]:
                arr[i] *= 2
                gained += arr[i]
                arr.pop(i + 1)
            i += 1
        arr += [0] * (4 - len(arr))
        return arr, gained

    @staticmethod
    def _rotate_right(g):
        return [[g[3 - c][r] for c in range(4)] for r in range(4)]

    def move(self, direction, sim=False):
        """direction: 0=left 1=right 2=up 3=down. Returns (moved, gained)."""
        rotations = {0: 0, 1: 2, 2: 3, 3: 1}
        g = [row[:] for row in self.grid]
        rots = rotations[direction]
        for _ in range(rots):
            g = self._rotate_right(g)

        moved = False
        total_gained = 0
        for r in range(4):
            new_row, gained = self._slide_row(g[r])
            if new_row != g[r]:
                moved = True
            g[r] = new_row
            total_gained += gained

        rots_back = (4 - rots) % 4
        for _ in range(rots_back):
            g = self._rotate_right(g)

        if moved:
            self.grid = g
            self.score += total_gained
            if not sim:
                if not self.won and self.max_tile() >= 2048:
                    self.won = True
                self.add_random()
                if not self.can_move():
                    self.over = True
        return moved, total_gained

    def can_move(self):
        if self.empty_cells():
            return True
        for r in range(4):
            for c in range(4):
                v = self.grid[r][c]
                if c < 3 and self.grid[r][c + 1] == v:
                    return True
                if r < 3 and self.grid[r + 1][c] == v:
                    return True
        return False

    def max_tile(self):
        return max(v for row in self.grid for v in row)

    def valid_moves(self):
        moves = []
        for d in range(4):
            b = self.clone()
            moved, _ = b.move(d, sim=True)
            if moved:
                moves.append(d)
        return moves
